src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs

```diff
@@ -483,17 +479,38 @@ private static void RemoveDirectoryInternal(DirectoryInfo directory, bool recurs
                     case Interop.Error.EPERM:
                     case Interop.Error.EROFS:
                     case Interop.Error.EISDIR:
-                        throw new IOException(SR.Format(SR.UnauthorizedAccess_IODenied_Path, directory.FullName)); // match Win32 exception
+                        throw new IOException(SR.Format(SR.UnauthorizedAccess_IODenied_Path, fullPath)); // match Win32 exception
                     case Interop.Error.ENOENT:
-                        if (!throwOnTopLevelDirectoryNotFound)
+                        if (!topLevel)
                         {
-                            return;
+                            return true;
+                        }
+                        goto default;
+                    case Interop.Error.ENOTDIR:
+                        if (topLevel)
+                        {
+                            if (!DirectoryExists(fullPath))
+                            {
+                                throw Interop.GetExceptionForIoErrno(Interop.Error.ENOENT.Info(), fullPath, isDirectory: true);
```

Comment:
@adamsitnik I would prefer to change this error code because I think the previous behavior was unintentional.
If you try to delete a socket file using `Directory.Delete`, the code used to throw `ENOENT`.
I think it is more appropriate to throw `ENOTDIR`. Do you agree?

Reply:
@adamsitnik can you share your opinion?